### ISSUE TITLE
Can't extract sprite to a custom location with a custom name

### ISSUE THREAD

**Issue by mistic** [kisenka/svg-sprite-loader#109]
Hi @kisenka ,

I expecting that I will be able to do anything like:

```javascript
{
  test: /\.svg(\?v=\d+\.\d+\.\d+)?$/,
  include,
  exclude,

  use: {
    loader: 'svg-sprite-loader',
    options: {
      symbolId: '[folder]-[name]-[hash:8]',
      extract: true,
      spriteFilename: 'assets/svg/sprite.[hash:8].svg',
    }
  },
},
```

However neither the sprite is emitted as 'sprite.svg' as I can't load the sprite icons at all, I'm always getting undefined.

**Comment by kisenka** [comment]
I am not sure that your loader notation is correct. `use` should be array of objects/strings. But when single loader is used simply use `loader` property:

```js
{
  test: /\.svg(\?v=\d+\.\d+\.\d+)?$/,
  include,
  exclude,
  loader: 'svg-sprite-loader',
  options: {
    symbolId: '[folder]-[name]-[hash:8]',
    extract: true,
    spriteFilename: 'assets/svg/sprite.[hash:8].svg',
  }
}
```

Also there is no `[hash]` token in spriteFilename, you should use `[spritehash]`. I think it's a bit confusing, it will be changes in next minor release.

**Comment by mistic** [comment]
Thanks for your answer @kisenka.

1 - Ok the hash thing is solved but is very awkward. It will be very easy just be able to use the custom tokens from `interpolateName`. What do u think @kisenka ?

2- I don't know why but this is the only loader i'm using that don't work with that loader notation:

```javascript
use: {
    loader: 'svg-sprite-loader',
    options: {
      symbolId: '[folder]-[name]-[hash:8]',
      extract: true,
      spriteFilename: 'assets/svg/sprite.[hash:8].svg',
    }
},
```

however using as u suggest worked.

3- Last but not least,  why I'm not able to use
```html
<svg>
    <use xlink:href="#ID" />
</svg>
```
with extract enabled? It only works if I use <img> tag.

**Comment by kisenka** [comment]
1. I agree, `[hash]` token is more convenient, as it's used in other interpolate places.
2. You're right, it's valid notation according to [webpack's json schema](https://github.com/webpack/webpack/blob/f2a6439b59edf1bbacdbfa3d7a702ee6f3671af8/schemas/webpackOptionsSchema.json#L690), sprite-loader should implement such notation recognition (but honestly it's minor priority :)
3. It's just not working in browsers at all, you can try it yourself without sprite-loader. I think it's because `<use>` tag should refer to existing node in DOM (even in shadow DOM).

**Comment by mistic** [comment]
1. Fab @kisenka 👍 
2. Yeah it isn't a show stopper, but I think u should add it to the roadmap as low priority?
3. Yeah that's the reason precisely. However how it will work if u want to load the svg sprite from a cdn for example? In my opinion it should be important having a way of always importing the generated sprite.svg into the DOM and be able to use `<svg><use></svg>`. The most easy way should be doing anything like that 

```javascript
(function(path, baseUrl) {
  var id = 'svg';
  var xhr = new XMLHttpRequest;
  var body = document.body;
  var div = document.createElement('div');
  var base = baseUrl || window.location.protocol + '//' + window.location.hostname + (window.location.port ? ':' + window.location.port : '');
  var url = base + path;

  if (div.id = id, body.insertBefore(div, body.childNodes[0]), 'withCredentials' in xhr) {
    xhr.withCredentials;
    xhr.open('GET', url, true);
  }
  else {
    if (typeof XDomainRequest == 'undefined') {
      return void(body.className += ' no-svg');
    }

    xhr = new XDomainRequest;
    xhr.open('GET', url);
  }

  xhr.onload = function() {
    div.className = 'u-visually-hidden';
    div.innerHTML = xhr.responseText;
  };

  xhr.onerror = function() {
    body.className += ' no-svg';
  };

  setTimeout(function() {
    xhr.send();
  }, 0);
})('/assets/svg/sprite.svg');
```

However my biggest doubt is how to know the name of the svg sprite after being hashed. Do u have any idea @kisenka ? Maybe u can export a global var in your webpack plugin after the emitting the svg sprite file like __SVG_PRITE__LOADER__SPRITE__NAME__ no? If u do that, passing this variable to the code I post above, we'll be able to load our sprite into the dom, fetching it from wherever we want (cdn or still our bundle server) and also simple use `<svg><use></svg>`. What do u think @kisenka ?

**Comment by kisenka** [comment]
2. Will do.
3. Why not use runtime loader mode? Extracting created mostly for importing SVGs from CSS.

**Comment by mistic** [comment]
@kisenka can u give me an example of your idea? I think I'm not getting it.

**Comment by kisenka** [comment]
I mean that default loader behaviour for images imported from JS is to generate sprite dinamycally at runtime - http://kisenka.me/svg-sprite-loader/examples/browser-sprite

**Comment by mistic** [comment]
@kisenka this is not a good idea :( In production builds we want to reduce the bundle size and be able to make Dynamic imports. In my opinion that's the true value of webpack. Doing what u are suggesting will increase the bundle size a lot because the spritesheet is always shipped with the js :( So it's really a good idea be able to extract the spritesheet but also load it from where we want (cdn or any other server). Besides that it's also cool being able to keep using `<svg><use></svg>` to be able to for example change some svg params of the image being used. What do u think @kisenka ?

**Comment by kisenka** [comment]
You can move spritesheet modules in separate chunk via CommonsChunkPlugin and download lazily, if you worry about the size. It will be nearly the same, plus small overhead as runtime module (it's 6kB gzipped). But we're using this technique at www.jetbrains.com - quite popular site and it's work fine. However it's nothing to talk about till we don't have any numbers about bundle size. It's just my thoughts.

> that is how to know the name of the svg sprite after being emitted with an hash.

That's simple, you can parse webpack compilation statistics and get any info about the build. It already done by guys from FormidableLabs, check [webpack-stats-plugin](https://github.com/FormidableLabs/webpack-stats-plugin). You can download this file at runtime, parse it and fetch any chunk via ajax.

**Comment by mistic** [comment]
1 - Can u explain better your first approach? Are u telling me to just wrap all my React components that are importing an svg image from the spritesheet in a seperate chunk? I don't like a lot this approach @kisenka.

2- @kisenka Just to inform u that I was able to load an image from the spritesheet using `<svg><use></svg>`. The difference is: when extracted we don't need to append #. We just need to use `<svg><use xlink:href="/assets/svg/sprite.a5a1af30.svg#login-warning-bd0f07e7-usage"></svg>`. However when we dont have extracted the spritesheet we'll need to append # to the [returnedObjectByImport].id because it will be simple "login-warning-bd0f07e7-usage" so it will be:
`<svg><use xlink:href="#login-warning-bd0f07e7-usage"></svg>`. 

3- However I still have a question for you @kisenka . If I want to serve this spritesheet over a CDN for example, it will be enough to set for example the publicPath on webpack and after the importing the svg image it will be returned with the publicPath url prepended?

**Comment by kisenka** [comment]
1. No, I suggest to move only sprite related module and load them lazily. I will create example to demonstrate.
2. Cool, it should be tested in all browsers, especially in Safari.
3. Yes, publicPath will work, but only when both files (page & sprite) are on the same domain, browsers will block cross-domain request.

**Comment by mistic** [comment]
1. Ok I'll wait for your example.

2. I've already tested and at least in the last version it's working. The only thing differing from using `<img>` tag or `<svg><use></svg>` tag is the files listed under Chrome Dev Tools. Using the `svg` one the folder containing the svg spritesheet is not listed:

<img width="267" alt="screen shot 2017-05-09 at 22 17 08" src="https://cloud.githubusercontent.com/assets/4440734/25873214/87a40650-3505-11e7-8b02-9fef31ecb6b8.png">

However the file is under assets/svg/sprite.[hash].svg and using `<img>` tag it appears. Do u have any idea why and also do u have any idea of an workaround for this @kisenka ? Also it's worth it apply efforts to the file appears listed on the sources in the chrome dev tool?

3. Yeah it should be possible using google cloud cdn :)

**Comment by kisenka** [comment]
2. Have no idea. BTW file appears in Network tab.

**Comment by mistic** [comment]
I'll wait for your example regarding to my point 1.

Anyway @kisenka , u don't think that it's a good idea saving a global variable with the name, or the names, of the generated spritesheets? In my opinion it's completely essencial! It can be disable by default and only defined if user wants. I found an example repo doing a thing very close:
https://github.com/mrsum/webpack-svgstore-plugin

At least in my head it makes a lot of sense due to the reasons I refer before. At the moment even with my last findings isn't possible to cover the use case at 100%. Also it's very easy to implement it in your plugin, so it's a win win :)

**Comment by rantiev** [comment]
What a stupid WEBPACKERS idea to connect any images and CSS from JS (that's only usefull if you have asynchronous on demand modules loading, that breaks classic way of loading css/images simultaneously with JS files). I understand that it needed sometime, but considering this as default way is wrong. 

On my end here i'm fcking with using sprite in HTML and CSS both. If i turn on extract mode i don't see how i can have them in HTML, if i use default way, i don't see how i can use them in CSS.

Ok. You connected your image from HTML + JS like this:
`import iconTwitter from '../../../imgs/twitter.svg';`
`	<svg>
		<use xlink:href="#twitter"></use>
	</svg>`

But how are you supposed to connect it with CSS then?

**Comment by kisenka** [comment]
@rantiev please be polite and respect others, if you don't want to be banned.

> But how are you supposed to connect it with CSS then?

In CSS you just refer to image as background-image. See extract example - [source](https://github.com/kisenka/svg-sprite-loader/blob/master/examples/extract-sprite/main.css) and [build](https://github.com/kisenka/svg-sprite-loader/blob/master/examples/extract-sprite/build/main.css)

**Comment by peterfoeng** [comment]
Hi @mistic and @kisenka

I am trying out the extract example but the html output is not hashed although webpack is building as hashed svg sprite, am I missing something here?

This is how I reference the sprite from the html:
```
  <h1>Sprite image embedded in entry.html (which processed by html-loader and extracted by extract-text-webpack-plugin)</h1>
    <svg role="img">
      <use xmlns:xlink="http://www.w3.org/1999/xlink" xlink:href="img/sprite.svg#icon-twitter"/>
    </svg>
```

part of my webpack:

```
 module: {
    rules: [
      {
        loader: 'svg-sprite-loader',
        test: /\.svg$/,
        options: {
          // override default options
          extract: true,
          spriteFilename: 'img/sprite.[hash:6].svg'
        },
      },
      {
        test: /\.css$/,
        loader: CSSExtractor.extract({ use: 'css-loader' })
      },
      {
        test: /\.html$/,
        loader: HTMLExtractor.extract({ use: 'html-loader' })
      }
    ]
  },
```

the html output is not hashed at all, any ideas how should I approach this? Thanks!

**Comment by kisenka** [comment]
@peterfoeng which version is used?

**Comment by peterfoeng** [comment]
@kisenka 3.0.1

**Comment by mistic** [comment]
@peterfoeng I didn't understand your problem, sorry :( Can u better explain the problem you're facing?

**Comment by kisenka** [comment]
@peterfoeng the best way - is to create a repo with reproducible behaviour :)

**Comment by peterfoeng** [comment]
Ignore me....I think I am using version 2.x something but the problem is resolved now that I am using 3.x something

**Comment by kisenka** [comment]
Closing this one be because:
- `[hash]` was implemented in [2.1.0](https://github.com/kisenka/svg-sprite-loader/releases/tag/v2.1.0).
- For object literal config notation see #118.
- For sprite stacking CSS and markup optional in extract mode see #132.
